pkg/security/podsecuritypolicy/provider.go

```diff
@@ -449,3 +450,13 @@ func validateRuntimeClassName(actual *string, validNames []string) field.ErrorLi
 	}
 	return field.ErrorList{field.Invalid(field.NewPath("spec", "runtimeClassName"), *actual, "")}
 }
+
+func allowsVolumeType(allowedVolumes sets.String, fsType policy.FSType) bool {
+	if allowedVolumes.Has(string(fsType)) {
+		return true
+	}
+	if allowedVolumes.Has(string(policy.Secret)) && fsType == policy.Projected && utilfeature.DefaultFeatureGate.Enabled(features.BoundServiceAccountTokenVolume) {
```

Comment:
We should just check for all types right?  That would match the intent of the PSP?  I guess we cannot check explicitly for SA token but I think `policy.Secret` is a close enough approximation.

```go
var projectedVolumeTypes = []string{ policy.Secret, policy.DownwardAPI, policy.ConfigMap }

if fsType == policy.Projected && allowedVolumes.HasAll(projectedVolumeTypes...) {
```

We could also make the check more specific by looking at `volume.Projected.Sources`.

Reply:
why we need to special casing the `DownwardAPI` and `ConfigMap` even though their projected ones are required for projected service account tokens?

the idea is to have a more smooth transition from secret based token to projected based token.

Reply:
I think we need to look at the contents of `volume.Projected.Sources` more carefully so that we can say that "because this PSP allows secret volumes, it is okay for it allow this projected volume which only mounts secrets or SA tokens" (i.e. we would reject a projected volume that uses `DownwardAPI` or `ConfigMap`).

Reply:
The projected service account volume will use a secret, configmap, and downwardAPI source. We could closely validate the specific shape of the projected sources to make sure they only pull in information equivalent to a service account token.

Reply:
> We could closely validate the specific shape of the projected sources to make sure they only pull in information equivalent to a service account token.

to minimize the blast radius, we can make a `reflect.DeepEqual` on the volume source with the one injected to pods https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/serviceaccount/admission.go#L528-L563.

PTAL

Reply:
> The projected service account volume will use a secret, configmap, and downwardAPI source

doh, my bad.  For some reason I was thinking the projected volume used for SA tokens had a different shape.